### ISSUE TITLE
dataset toString variants

### ISSUE THREAD

**Issue by genmeblog** [techascent/tech.ml.dataset#78]
By default printing dataset results in markdown table - which is perfectly ok. But when dataset includes another dataset, printing returns mess.

Could it be possible to add an option (maybe in meta data?) to return just first part (name and shape) when this option is set? This could work when meta contained `:to-string :minimal` for example.

What do you think?

**Comment by cnuernber** [comment]
I think that there are a few things going on here:
1.  The print-table code in clojure.pprint fails with any entry that tostrings to multiple lines.  So the fix for this is probably just trimming the strings to be the first line and adding an elipsis to make it clear the input is edited.  This fixes tensors, datasets, lots of things.
2.  I think extending/rewriting the print-table code to handle multiple line strings is also wise and there is precedent to just measure the line and add text as necessary.  But this likely isn't going to be what you want as a default, in most cases the one-line pathway will actually be what you want.  Then if you do want to embed this in markdown you can enable multi-line either globally or with metadata.
3.  I wonder, specifically if we are going to get into it, what the options are for markdown tables.

**Comment by genmeblog** [comment]
Ad.1/2. Trimming is good solution. But this requires `print-table` to know the depth.

Ad.3. I'm testing currently pandoc (through knitr/rmarkdown) and here we have what is possible (take a look at `grid_tables` and `multiline_tables`): https://pandoc.org/MANUAL.html#tables, pandoc renders to github markdown so it should be ok. `print-tables` generates `pipe_tables` variant.
With `grid_tables` you can put another table inside the cell, but I'm getting the error converting such case to PDF (HTML works).

So to summarise: trimming is possibly the best now.

**Comment by genmeblog** [comment]
One more comment, Clojure variant doesn't render in Github Markdown by default because of `+`  (plus) usage in header delimiter instead of `|` (pipe)

Markdown `pipe_tables`, default variant

```md
| a | b |
|---|---|
| 1 | 2 |
```

| a | b |
|---|---|
| 1 | 2 |

Markdown `pipe_tables`, Clojure variant

```md
| a | b |
|---+---|
| 1 | 2 |
```

| a | b |
|---+---|
| 1 | 2 |

**Comment by cnuernber** [comment]
OK, then I think that settles it.  I am currently using clojure.pprint/print-table.  If I have to get into it to fix the `+` symbol I can start building out the multiline variants.

**Comment by cnuernber** [comment]
This is what I have so far:
https://gist.github.com/cnuernber/7c99b8f559ffca146181e4c85d4a6fb3

**Comment by genmeblog** [comment]
That's cool!

**Comment by cnuernber** [comment]
New release is up and documentation is in the [walkthough](https://github.com/techascent/tech.ml.dataset/blob/master/docs/walkthrough.md#printing).

**Comment by genmeblog** [comment]
@cnuernber another small fix is needed. There should be empty line before table starts. Current solution is ok for github markdown but `pandoc` is lost and prints just text.

The following...

```
_unnamed [2 2]:
| :a |                            :b |
|----|-------------------------------|
|  1 | #tech.v2.tensor<float64>[3 3] |
|    | [[0.000 1.000 2.000]          |
|    |  [3.000 4.000 5.000]          |
|    |  [6.000 7.000 8.000]]         |
|  2 | #tech.v2.tensor<float64>[3 3] |
|    | [[0.000 1.000 2.000]          |
|    |  [3.000 4.000 5.000]          |
|    |  [6.000 7.000 8.000]]         |

_unnamed [2 2]:

| :a |                            :b |
|----|-------------------------------|
|  1 | #tech.v2.tensor<float64>[3 3] |
|    | [[0.000 1.000 2.000]          |
|    |  [3.000 4.000 5.000]          |
|    |  [6.000 7.000 8.000]]         |
|  2 | #tech.v2.tensor<float64>[3 3] |
|    | [[0.000 1.000 2.000]          |
|    |  [3.000 4.000 5.000]          |
|    |  [6.000 7.000 8.000]]         |
```

...renders as:

![image](https://user-images.githubusercontent.com/38646601/82701563-b7137f80-9c70-11ea-8f24-b3905a3f1fce.png)

**Comment by cnuernber** [comment]
Ahh, Rich had that in the original and I thought it was just noise.  Will add back in.

**Comment by cnuernber** [comment]
OK, 49 is up.

**Comment by genmeblog** [comment]
Yay! Thanks, checking.

**Comment by genmeblog** [comment]
Another case found, when missing values are at the beginning or the end, they are stripped

```clojure
(ds/->dataset [{:a nil} {:a 2} {:a nil} {:a nil}])
;; => _unnamed [4 1]:
;;    | :a |
;;    |----|
;;    |  2 |
```

**Comment by genmeblog** [comment]
Also, empty dataset renders as:

```clojure
_unnamed [0 0]:

|  |
|--|
```

I think just name with shape is enough.

**Comment by cnuernber** [comment]
yep, thanks.  Nice touch.
